src/libraries/Microsoft.Extensions.Logging.Console/src/WindowsLogConsole.cs

```diff
@@ -17,42 +17,108 @@ public WindowsLogConsole(bool stdErr = false)
+                return true;
             }
+            return false;
+        }
 
+        public bool SetForegroundColor(ConsoleColor? foreground)
+        {
             if (foreground.HasValue)
             {
                 System.Console.ForegroundColor = foreground.Value;
+                return true;
             }
-
-            return background.HasValue || foreground.HasValue;
+            return false;
         }
 
         private void ResetColor()
         {
             System.Console.ResetColor();
         }
 
-        public void Write(string message, ConsoleColor? background, ConsoleColor? foreground)
+        public void Write(ReadOnlySpan<char> span, ConsoleColor? background, ConsoleColor? foreground)
         {
-            bool colorChanged = SetColor(background, foreground);
-            _textWriter.Write(message);
+            var colorChanged = SetColor(background, foreground);
+            _textWriter.Write(span);
             if (colorChanged)
             {
                 ResetColor();
             }
         }
 
-        public void WriteLine(string message, ConsoleColor? background, ConsoleColor? foreground)
+        public void Write(string message)
         {
-            bool colorChanged = SetColor(background, foreground);
-            _textWriter.WriteLine(message);
-            if (colorChanged)
+            // writes content with the embedded foreground and background ansi colored. it skips non-color related ansi codes
+            (int startIndex, int length, ConsoleColor? bg, ConsoleColor? fg) content = (-1, 0, null, null);
+            var span = message.AsSpan();
+            const char EscapeChar = '\x1B';
+            ConsoleColor? color = null;
+            bool isDarkColor = true;
+            for (int i = 0; i < span.Length; i++)
             {
-                ResetColor();
+                if (span[i] != EscapeChar || span.Length < i + 3 || span[i + 1] != '[')
+                {
+                    // not an escape char
+                    content.length++;
+                    if (content.startIndex == -1)
+                    {
+                        content.startIndex = i;
+                    }
+                }
+                else if (span[i + 3] == 'm')
+                {
+                    if (int.TryParse(span.Slice(i + 2, length: 1), out int escapeCode))
+                    {
+                        if (escapeCode == 1)
+                            isDarkColor = false;
+                        // parsing only ansi color codes
+                        i += 3;
+                    }
+                }
+                else if (span[i + 4] == 'm')
+                {
+                    if (int.TryParse(span.Slice(i + 2, length: 2), out int escapeCode))
+                    {
+                        if (SetsForegroundColor(escapeCode, isDarkColor, out color))
+                        {
+                            if (content.startIndex != -1)
```

Comment:
>content [](start = 32, length = 7)

do we care checking if the color is same as content.fg? same question apply on the background color too.